bigchaindb_driver/connection.py

```diff
@@ -25,36 +31,90 @@ def __init__(self, *, node_url, headers=None):
         if headers:
             self.session.headers.update(headers)
 
-    def request(self, method, timeout=None, *, path=None, json=None,
-                params=None, headers=None, **kwargs):
-        """Performs an HTTP requests for the specified arguments.
+        self._retries = 0
+        self.backoff_time = None
+
+    def request(self, method, *, path=None, json=None,
+                params=None, headers=None, timeout=None, **kwargs):
+        """Performs an HTTP request with the given parameters.
+
+           Implements exponential backoff.
+
+           If `ConnectionError` occurs, a timestamp equal to now +
+           the default delay (`BACKOFF_DELAY`) is assigned to the object.
+           The timestamp is in UTC. Next time the function is called, it either
+           waits till the timestamp is passed or raises `TimeoutError`.
+
+           If `ConnectionError` occurs two or more times in a row,
+           the retry count is incremented and the new timestamp is calculated
+           as now + the default delay multiplied by two to the power of the
+           number of retries.
+
+           If a request is successful, the backoff timestamp is removed,
+           the retry count is back to zero.
 
         Args:
             method (str): HTTP method (e.g.: ``'GET'``).
             path (str): API endpoint path (e.g.: ``'/transactions'``).
             json (dict): JSON data to send along with the request.
-            params (dict)): Dictionary of URL (query) parameters.
+            params (dict): Dictionary of URL (query) parameters.
             headers (dict): Optional headers to pass to the request.
+            timeout (int): Optional timeout in seconds.
             kwargs: Optional keyword arguments.
 
         """
-        url = self.node_url + path if path else self.node_url
-        response = self.session.request(
-            method=method,
-            timeout=timeout,
-            url=url,
-            json=json,
-            params=params,
-            headers=headers,
-            **kwargs
-        )
+        backoff_timedelta = self.get_backoff_timedelta()
+
+        if timeout is not None and timeout < backoff_timedelta:
+            raise TimeoutError
+
+        if backoff_timedelta > 0:
+            time.sleep(backoff_timedelta)
+
+        connExc = None
```

Comment:
what about `connExc = True` ? then we don't need `connExc is None`

Reply:
I didn't get you are trying to imply here. What I was trying to say is that if `connExc` is `None` is always converted to `True` we could rather just assign it a default value of `True`. You can choose either way

Reply:
`connExc` represents an exception - that's why I do not want to give it a default value of `True` since `True` is not really a value of an exception.